### PR TITLE
Recurring expenses

### PR DESCRIPTION
**Pull request by xinmingzh**
Implement list recurrence and add recurrence features. 

Adding a recurrence creates and adds multiple expenses to expenses in model. At the same time it also creates a recurrence contain a list of the expenses added, adds to the recurrence list in model.

Add recurrence does not save recurrences to json file, so recurrence info will be lost upon restart of application. Expenses will remain in the model.

Add recurrence currently does not support undo and redo.

@waynefong0401 Regarding the concern of adding multiple future expenses affecting statistics, I feel that it should be the same as manually adding future expense, so it is not a problem arising from recurrence feature. A possibility will be to have FutureExpense instances that does not show up in the user interface.

I am still adding more code so you guys don't need to help me merge.

### CODE REVIEW

**Comment by xinmingzh** [comment]
Test for new classes not added yet

**Comment by xinmingzh** [comment]
This version is should be okay to merge

File: docs/UserGuide.adoc [hunk 1]
```diff
@@ -129,6 +128,76 @@ Displays below expenses (for example) in the GUI:
 image::ListCommandExample.png[]
 {nbsp} +
 
+<<<<<<< HEAD
```
Comment (Davidcwh) [inline docs/UserGuide.adoc]: remove the `<<<<<<< HEAD`

File: docs/UserGuide.adoc [hunk 2]
```diff
@@ -146,6 +215,7 @@ This _actually_ deletes all your commands, use with caution! The undo feature ma
 
 [NOTE]
 Archived expenses are not cleared. +
+>>>>>>> master
```
Comment (Davidcwh) [inline docs/UserGuide.adoc]: Remove this also

File: src/main/java/seedu/billboard/logic/commands/AddRecurrenceCommand.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    private Name name;
+    private Description description;
+    private Amount amount;
+    private CreatedDateTime date;
+    private List<String> tagList;
+    private DateInterval interval;
+    private int iterations;
+
+    /**
+     * Creates an AddCommand to add the specified {@code Expense}
```
Comment (Davidcwh) [inline src/main/java/seedu/billboard/logic/commands/AddRecurrenceCommand.java]: Javadocs error `AddCommand` -> `AddRecurrenceCommand`

File: src/main/java/seedu/billboard/model/expense/Expense.java [hunk 4]
```diff
@@ -94,7 +94,12 @@ public void archiveTo(String archiveName) {
      * Returns deep clone of the expenses by creating a new Expense object with the same data.
      */
     public Expense getClone() {
-        return new Expense(this.name, this.description, this.amount, this.created, this.tags, this.archiveName);
+        return new Expense(
+                name.getClone(),
+                description.getClone(),
+                amount.getClone(),
+                created.getClone(),
+                new HashSet<>(tags));
```
Comment (Davidcwh) [inline src/main/java/seedu/billboard/model/expense/Expense.java]: left out archive name when making new object of expense!

File: src/main/java/seedu/billboard/model/recurrence/RecurrenceList.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+                }
+            }
+        }
+        return true;
+    }
+
+    public RecurrenceList getClone() {
+        RecurrenceList clonedList = new RecurrenceList();
+        for (Recurrence e : this.internalList) {
+            clonedList.add(e);
```
Comment (Davidcwh) [inline src/main/java/seedu/billboard/model/recurrence/RecurrenceList.java]: This is not actually making deep copy of `Recurrence` objects, need to initialize a new instance of `Recurrence` with the same values

File: src/main/java/seedu/billboard/model/expense/ExpenseList.java [hunk 6]
```diff
@@ -137,4 +165,8 @@ public ExpenseList getClone() {
         }
         return clonedList;
     }
+
+    public Expense get(int i) {
+        return internalList.get(0);
```
Comment (Davidcwh) [inline src/main/java/seedu/billboard/model/expense/ExpenseList.java]: Why get the 0 index expense?

File: src/main/java/seedu/billboard/model/recurrence/Recurrence.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+import seedu.billboard.model.expense.Amount;
+import seedu.billboard.model.expense.CreatedDateTime;
+import seedu.billboard.model.expense.Description;
+import seedu.billboard.model.expense.Expense;
+import seedu.billboard.model.expense.ExpenseList;
+import seedu.billboard.model.expense.Name;
+import seedu.billboard.model.tag.Tag;
+
+/**
+ * Represents a recurrence in the address book.
```
Comment (Davidcwh) [inline src/main/java/seedu/billboard/model/recurrence/Recurrence.java]: recurrence in Billboard**

File: src/main/java/seedu/billboard/logic/parser/AddRecurrenceCommandParser.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+import seedu.billboard.commons.core.date.DateRange;
+import seedu.billboard.logic.commands.AddRecurrenceCommand;
+import seedu.billboard.logic.parser.exceptions.ParseException;
+import seedu.billboard.model.expense.Amount;
+import seedu.billboard.model.expense.CreatedDateTime;
+import seedu.billboard.model.expense.Description;
+import seedu.billboard.model.expense.Name;
+
+/**
+ * Parses input arguments and creates a new AddCommand object
```
[the remark this answers is not in the document]
Comment (xinmingzh) [reply src/main/java/seedu/billboard/logic/parser/AddRecurrenceCommandParser.java]: Editted

File: src/main/java/seedu/billboard/logic/commands/AddRecurrenceCommand.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+import seedu.billboard.model.expense.CreatedDateTime;
+import seedu.billboard.model.expense.Description;
+import seedu.billboard.model.expense.Expense;
+import seedu.billboard.model.expense.ExpenseList;
+import seedu.billboard.model.expense.Name;
+import seedu.billboard.model.recurrence.Recurrence;
+import seedu.billboard.model.tag.Tag;
+
+/**
+ * Adds an expense to the address book.
```
Comment (Davidcwh) [inline src/main/java/seedu/billboard/logic/commands/AddRecurrenceCommand.java]: to the Billboard**

File: src/main/java/seedu/billboard/model/tag/Tag.java [hunk 10]
```diff
@@ -32,6 +32,10 @@ public static boolean isValidTagName(String test) {
         return test.matches(VALIDATION_REGEX);
     }
 
+    public String getTagName() {
+        return tagName;
+    }
+
```
Comment (waifonglee) [inline src/main/java/seedu/billboard/model/tag/Tag.java]: isnt tagName public?

Comment (xinmingzh) [reply src/main/java/seedu/billboard/model/tag/Tag.java]: I removed the method and changed my code access the variable directly.

File: src/main/java/seedu/billboard/commons/core/date/DateInterval.java [hunk 11]
```diff
@@ -17,6 +17,9 @@
     MONTH("month", Period.ofMonths(1), TemporalAdjusters.firstDayOfMonth()),
     YEAR("year", Period.ofYears(1), TemporalAdjusters.firstDayOfYear());
 
+    public static final String MESSAGE_CONSTRAINTS =
```
Comment (weiijiie) [inline src/main/java/seedu/billboard/commons/core/date/DateInterval.java]: This constant seems to not be used at all

File: src/main/java/seedu/billboard/commons/core/date/DateInterval.java [hunk 12]
```diff
@@ -46,6 +49,18 @@
         return Optional.ofNullable(nameToDateIntervalMap.get(name));
     }
 
+    /**
+     * Checks that {@code String interval} is a valid {@code DateInterval}
+     */
+    public static boolean isValidDateInterval(String interval) {
```
Comment (weiijiie) [inline src/main/java/seedu/billboard/commons/core/date/DateInterval.java]: Doesn't the function `intervalFromName` achieve the functionality of this method + the static method defined in ParserUtil? Also Enum.valueOf() only accepts the exact string representation of the enum, aka "MONTH", "DAY" and so on. Is that intended behavior, to get the user to input the intervals in capital letters?

File: src/main/java/seedu/billboard/commons/core/date/DateInterval.java (hunk 11 above)
Comment (xinmingzh) [reply src/main/java/seedu/billboard/commons/core/date/DateInterval.java]: It is used in the parseInterval() method in ParserUtil

File: src/main/java/seedu/billboard/commons/core/date/DateInterval.java (hunk 12 above)
Comment (xinmingzh) [reply src/main/java/seedu/billboard/commons/core/date/DateInterval.java]: The user is able to input intervals in both lower and upper case as the parseInterval() method converts the input to uppercase before calling Interval.valueOf()

**Pull request merged**